content/blog/posts/2022-08-04-add-native-notifications.md

```diff
@@ -0,0 +1,172 @@
+
+## What we'll work on today
+
+Let's create a basic notification that checks for and then gets the user's permission. Then we can show a notification about our cool new blog posts. We can then add some actions to allow the user to opt it future notifications as to when new posts are published. There's more to it, but we'll get to the nuances shortly.
+
+For this I'm going to create a singleton pattern file since I just need it to load and run when loaded. I find this to be a simple and readable pattern for our needs.
+
+```js
+class NotificationService {
+  constructor() {
+    // whatever you want once instance is created
+  }
+
+  async _isShowNotificationPermitted() {
+    // ...
+  }
+
+  async _showNotification() {
+    // ...
+  }
+}
+
+new NotificationService();
+```
+
+Using this javascript class I am adding a couple methods to handle checking the permission settings, requesting permissions, and then showing a notification if granted. Let's start with checking for and asking for permission to show notifications from our site.  
+
+```js
+async _isShowNotificationPermitted() {
+  if ('Notification' in window) {
+    let { permission } = Notification;
+    if (permission !== 'granted') {
+      permission =  await Notification.requestPermission();
+    }
+    if (permission !== 'denied') {
+      return true;
+    }
+  }
+
+  return false;
+}
+```
+
+I'll walk through the method a bit. First we're just basically checking for support in the browser in the off chance your user doesn't have a supported browser, otherwise why have those needless errors? The default value from `Notification.permission` is just that, `default`, which we treat essentially the same as `denied`. In that case we trigger `requestPermission()` in the API to request access from the user. If the user approves, we can move on to showing them notifications until this permission is revoked. Let's look then at a method for showing one.
+
+```js
+async _showNotification() {
+  const permission = await this._isShowNotificationPermitted();
+
+  if (permission !== 'denied') {
+    new Notification('Be notified of new posts', {
+      body: 'Let us notifiy you of new posts on our site!',
+    });
+  }
+}
+```
+
+So this one is a bit more straightforward. We check permission as noted in the previous code example, then as long as we're not denied then a notification can be shown. There's really only one required argument, a title. I'm adding a basic option for body text here as well. This will just show a notification and the user can dismiss it or select to go to browser settings if they'd like to change the permissions. Let's make sure our instance fires off when loaded before we move on.
```

Comment:
```suggestion
This one is a bit more straightforward. We check permission, as noted in the previous code example, and as long as we're not denied, a notification can be shown. There's really only one required argument, a title. Here we are adding a basic option for body text as well. This will just show a notification and the user can dismiss it or select to go to browser settings, if they'd like to change the permissions. Let's make sure our instance fires off when loaded before we move on.
```